consensus/consensus.go

```diff
@@ -85,8 +83,58 @@ func (fc *FastHotStuffEventLoop) Init(
 	return nil
 }
 
-func (fc *FastHotStuffEventLoop) UpdateChainTip( /* TODO */ ) {
-	// TODO
+// ProcessSafeBlock must only be called when the caller has accepted a new block, connected it
+// to the tip of the blockchain, and determined that the block is safe to vote on. Given such a
+// block, this function resets the internal timers and state of the Fast HotStuff consensus that
+// determine the next action. The functions expects the following for the input params:
+//   - block: the input block that was safely added to the blockchain and is safe to vote on
+//   - validators: the validator set for the next block height
+func (fc *FastHotStuffEventLoop) ProcessSafeBlock(block Block, validators []Validator) error {
+	// Grab the consensus instance's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the consensus instance is running
+	if fc.status != consensusStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessSafeBlock: Consensus instance is not running")
+	}
+
+	// Do a basic integrity check on the block
+	if !isProperlyFormedBlock(block) {
+		return errors.New("FastHotStuffEventLoop.ProcessSafeBlock: Invalid block")
+	}
+
+	// Do a basic integrity check on the validator set
+	if !isProperlyFormedValidatorSet(validators) {
+		return errors.New("FastHotStuffEventLoop.ProcessSafeBlock: Invalid validator set")
+	}
+
+	// Update the chain tip and validator set
+	fc.chainTip = block
+
+	// We track the current view here so we know which view to start the timeout timer for.
+	fc.currentView = block.GetView() + 1
```

Comment:
I think you should make currentView be +1 of the QC or the AggregateQC in the block. Check out [AdvanceView in the reference](https://github.com/deso-protocol/hotstuff_pseudocode/blob/1c5b4f085f89efa1984741d4bd174ce90190d98a/fast_hotstuff_bls.go#L224).

Reply:
As discussed offline, the `fc.currentView = block.GetView() + 1` is correct as we're voting on the view the block was proposed.